File: src/main/java/com/github/joelws/release/tracker/service/artifact/operation/DeleteArtifactServiceOperation.java
package com.github.joelws.release.tracker.service.artifact.operation;


import com.github.joelws.release.tracker.conversion.ArtifactDtoToArtifactAdapter;
import com.github.joelws.release.tracker.dto.artifact.ArtifactDto;
import com.github.joelws.release.tracker.entity.artifact.Artifact;
import com.github.joelws.release.tracker.service.ServiceHelper;
import com.github.joelws.release.tracker.service.ServiceOperation;
import com.github.joelws.release.tracker.service.artifact.execution.DeleteArtifactServiceExecution;

import javax.ws.rs.core.Response;

public class DeleteArtifactServiceOperation extends ServiceOperation<String> {

    private final ServiceHelper helper;

    private final DeleteArtifactServiceExecution deleteArtifactServiceExecution;

    public DeleteArtifactServiceOperation(final ServiceHelper helper,
                                          final DeleteArtifactServiceExecution deleteArtifactServiceExecution) {
        this.helper = helper;
        this.deleteArtifactServiceExecution = deleteArtifactServiceExecution;
    }

    @Override
    protected Response delegate(String json) {

        ArtifactDtoToArtifactAdapter artifactDtoToArtifactAdapter = helper.getFactory().getImpl(ArtifactDtoToArtifactAdapter.class);

        Artifact artifactInQuestion = artifactDtoToArtifactAdapter.adapt(helper.getJsonAdapter().getObjectFromJson(json, ArtifactDto.class));

        deleteArtifactServiceExecution.execute(artifactInQuestion);

        return helper.getRestResponseBuilder().build(200);
    }
}

File: src/main/kotlin/com/github/joelws/release/tracker/service/artifact/operation/DeleteArtifactServiceOperation.kt
package com.github.joelws.release.tracker.service.artifact.operation

import com.github.joelws.release.tracker.conversion.ArtifactModelAdapter
import com.github.joelws.release.tracker.entity.artifact.Artifact
import com.github.joelws.release.tracker.interfaces.Adapter
import com.github.joelws.release.tracker.model.artifact.ArtifactModel
import com.github.joelws.release.tracker.response.RestResponse.Success
import com.github.joelws.release.tracker.response.build
import com.github.joelws.release.tracker.service.ServiceExecution
import com.github.joelws.release.tracker.service.ServiceHelper
import com.github.joelws.release.tracker.service.ServiceOperation
import javax.ws.rs.core.Response

open class DeleteArtifactServiceOperation(val helper: ServiceHelper,
                                          val deleteArtifactServiceExecution: ServiceExecution<Artifact, Unit>) : ServiceOperation<String> {

    override fun delegate(param: String?): Response {

        @Suppress("UNCHECKED_CAST")
        val artifactModelAdapter: Adapter<ArtifactModel, Artifact> = helper.adapterFactory.getAdapter(ArtifactModelAdapter::class.java) as Adapter<ArtifactModel, Artifact>

        val artifactInQuestion = artifactModelAdapter.adapt(helper.jsonAdapter.getObjectFromJson(param, ArtifactModel::class.java))

        deleteArtifactServiceExecution.execute(artifactInQuestion)

        return Success().build()

    }


}
